sycl/doc/design/Fuzzing.md

```diff
@@ -0,0 +1,104 @@
+variable should be fuzzed.
+
+The most interesting are ones which expect data in a certain format, like
+`ONEAPI_DEVICE_SELECTOR` or `SYCL_CACHE_TRESHOLD`.
+
+[sycl-rt-env-variables]: https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md
+
+### Fuzzing sycl config file
+
+Instead of tweaking SYCL RT behavior through environment variables, the same
+can be done by providing a config file. We don't seem to have any documentation
+on it except the source code for the functionality which is located in
+`sycl/source/detail/config.cpp` file.
+
+There is a prototype for the sycl config file fuzzer available at
+https://github.com/intel/llvm/pull/16308
+
+### Fuzzing API entry points
+
+TODO: think more about this section, i.e. whether or not we want to fuzz SYCL
+APIs. Not every of them accepts "raw" data, but instead expects some SYCL
+objects returned from previous API calls. However, there are still plenty of
+APIs which accept raw pointers and other fundamental data types. Note: to
+properly fuzz them structure-aware fuzzing may be needed.
+
+## SYCL Compiler
+
+SYCL compiler is based on the upstream LLVM and it is an enormously huge
+codebase. Some of LLVM components have been re-used without any modifications to
+them at all. Some of LLVM components were slightly tweaked or significantly
+modified and there are components which are completely new and only exist in
+our implementation.
+
+For every re-used component we should be able to benefit from existing fuzz
+testing written for those. Upstream documentation has them documented
+[here][llvm-fuzzers].
+
+[llvm-fuzzers]: https://llvm.org/docs/FuzzingLLVM.html
+
+However, even though we could re-use existing fuzzers, we can't just rely on
+someone else running them on the upstream codebase, because those runs won't
+cover any customizations we made (including new components like optimization
+passes which we added only in our downstream).
+
+There are also some unique components which may require special fuzzers.
+Sections below will go through components that we have and describe in more
+details like what should we fuzz and if we already have an existing fuzzer for
+that.
+
+There is also [intel/yarpgen](https://github.com/intel/yarpgen) project that can
+be used to fuzz SYCL compilers. It generate random programs (of certain
+structure) to detect weaknesses and bugs in optimization passes.
+
+### Command line options
+
+There are plenty of SYCL-specific command line options and there are multiple of
+those which are not mere flag, but expect a user-provided value in a certain
+format.
+
+Those options should be fuzzed as well to ensure proper error handling of
+various weird inputs.
+
+### SYCL-specific passes
+
+We have developed a number of passes to implement different SYCL features. They
+all can be found in `llvm/lib/SYCLLowerIR` folder. We don't need a dedicated
+fuzzer for every pass, we can re-use existing LLVM fuzzer intended for compiler
```

Comment:
```suggestion
fuzzer for every pass, but can instead re-use existing LLVM fuzzer intended for compiler
```

Reply:
Applied in db6aba9154a2989f998a7c77ac8e388160dd9b18